### ISSUE TITLE
Feature request: generalize 'extract_yyyy_mm_dd_hh_mm_ss_from_datetime64' to accept array arguments

### ISSUE THREAD

**Issue by DanJonesOcean** [ECCO-GROUP/ECCOv4-py#103]
Hello,

At present, the 'extract_yyyy_mm_dd_hh_mm_ss_from_datetime64' function only accepts scalar arguments. If you give it an array of datetime64 values, it returns an error. Here is some code that illustrates the error:

```
# import  modules
import numpy as np
import xarray as xr
import xgcm
import sys

# import ecco_v4_py
sys.path.append('/Users/USERNAME/ECCOv4-py/ECCOv4-py')
import ecco_v4_py as ecco

# define main directory
base_dir = '/Users/USERNAME/ECCOv4-py/Version4/'
# define ECCO version
ecco_version = 'v4r4'
# define a high-level directory for ECCO fields
ECCO_dir = base_dir + '/Release4'
# define data directory
data_dir= ECCO_dir + '/nctiles_monthly'

# load monthly mean files
ds = ecco.recursive_load_ecco_var_from_years_nc(data_dir, \
                                                vars_to_load=['THETA'],\
                                                years_to_load=range(1993, 2017))

# extract xarray DataArray (an array of datetime64[ns] values)
tds = ds.time
```
Running the 'extract' function on a single scalar datetime64[ns] works fine:

```
ecco.extract_yyyy_mm_dd_hh_mm_ss_from_datetime64(ds.time[-1].values)
(2016, 12, 16, 12, 0, 0)
```
But giving the same function an array of datetime64[ns] values as follows

```
ecco.extract_yyyy_mm_dd_hh_mm_ss_from_datetime64(ds.time.values)
```
produces this error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-8faeaf0a92b0> in <module>()
----> 1 ecco.extract_yyyy_mm_dd_hh_mm_ss_from_datetime64(ds.time.values)

/Users/USERNAME/ECCOv4-py/ECCOv4-py/ecco_v4_py/ecco_utils.py in extract_yyyy_mm_dd_hh_mm_ss_from_datetime64(dt64)
    189 
    190     s = str(dt64)
--> 191     year = int(s[0:4])
    192     mon = int(s[5:7])
    193     day = int(s[8:10])

ValueError: invalid literal for int() with base 10: "['19"
```

Could the function `extract_yyyy_mm_dd_hh_mm_ss_from_datetime64` in `ecco_utils` be generalized to accept array arguments as well?

**Comment by DanJonesOcean** [comment]
In case it's useful, here is my conda environment:

[https://anaconda.org/dannes/eccov4](https://anaconda.org/dannes/eccov4)

**Comment by timothyas** [comment]
Hey @DanJonesOcean thanks for raising this! Yes it should totally be possible to work with arrays. Would something like this work?

```
yarr,marr,darr,harr,mins,secs = [ecco.extract_yyyy_mm_dd_hh_mm_ss_from_datetime64(x) for x in ds.time.values]
```

We could also test for arrays in the function. However, I'm not really convinced this would be necessary (open to being convinced though!), because if you want an array with the months/days/years, this is already baked into xarray. For example, you can get xarray DataArrays with the years, month, day, etc:

```
yrs = ds.time.dt.year
months = ds.time.dt.month
```

and so on for `day`, `hour`, `minute`, `second`, ... and even more options [here](http://xarray.pydata.org/en/stable/time-series.html#datetime-components).

I hope that helps! Let me know what you think.

**Comment by DanJonesOcean** [comment]
Hi @timothyas! That would probably work okay, yes. I ran into that error because I was trying to adapt an existing example to another use case, and I was surprised that it didn't work when I passed it an array. 

If you don't think it's needed, that's okay. Could I instead request that the function do some input checking to make sure that it has only been passed a scalar? Maybe it could return an error in that case. What do you think?

**Comment by timothyas** [comment]
Hey @DanJonesOcean, sorry for the delay - lots of spinning plates! So yes I think we could test the inputs and raise an error. Or actually perhaps it is better to simply handle array inputs. I'm actually tending toward this latter option now that I'm thinking about it. Since the function is available, it might as well be flexible, since I don't think it would be hard to do... (famous last words).

We should additionally test for this in our test suite (which is... ridiculously small right now...), since it is an available function. We can add this functionality and test when we get a chance, but of course if you feel comfortable and have the time, contributions via PRs are always welcome :)

**Comment by timothyas** [comment]
Hey @DanJonesOcean! I was able to get to this and expanded the functionality of this function. Now it can handle xarray DataArrays, numpy arrays, lists, and of course a single date, where you get what you give (there's a Beatles quote in there somewhere).

For now you can get this by cloning the repo or via pip:
```
pip install git+https://github.com/ECCO-GROUP/ECCOv4-py.git
```

I am adding some new features soon, and then make another release. Then this will be available through the usual pip and conda-forge methods as well.

**Comment by DanJonesOcean** [comment]
Thanks, @timothyas! That's exciting.  : )

**Comment by DanJonesOcean** [comment]
If you're happy, it sounds like we can close this issue. Would you like to do the honors?

**Comment by timothyas** [comment]
Awesome! I'll actually go ahead and leave it open for now until a new version is released, which I think will be after I merge some new stuff soon...

**Comment by ifenty** [comment]
@timothyas I think we can close this now?

**Comment by timothyas** [comment]
Whoops, yes this is all taken care of. Sorry I let this slip by. This function, along with generalizations of a lot of the calculations (computing volume transport etc) are available with ecco_v4_py version 1.4.0 which is available on conda-forge and pip. Cheers 🚀
